### PR TITLE
 :seedling: KAI evaluated data Visualizer

### PR DESCRIPTION
**Pull request by midays**
Solves [Issue #15](https://github.com/konveyor/kai-ci/issues/15)

This script takes the KAI Evaluation data in the form of a json an creates 2 Main chart sections:

- The first chart section shows the performance history for KAI nightly runs for a selected date range
- The second charts section is for a details data represtnation of a single KAI run, a pie chart that shows the valid/ invalid code, and the necessary/unnecessary changes. The second chart is a stacked chart consisting of a bar chart that shows the range of the for fields ie. `['Effectiveness', 'Specificity', 'Reasoning', 'Competency']` and a line chart that shows the average value for each category.

screencapture-file-home-midays-Documents-projects-kai-ci-kai-client-ci-report-index-html-2024-11-05-16_51_47

### CODE REVIEW

File: kai-client-ci/report/index.html [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Dynamic Chart from JSON Data</title>
+    <script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
+    <link rel="stylesheet" href="index.css">
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/flatpickr/dist/flatpickr.min.css">
+    <script src="https://cdn.jsdelivr.net/npm/flatpickr"></script>
+    <style>
```
Comment (abrugaro) [inline kai-client-ci/report/index.html:12]: Remove empty tag

File: kai-client-ci/report/index.html [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                <input type="date" id="date-picker" />
+            </div>
+            <div id="single-performace-charts">
+                <canvas id="pies-chart"></canvas>
+                <canvas id="average-ranges-chart"></canvas>
+            </div>
+        </div>
+    </div>
+    <div id="footer"></div>
+    <script src="./visualizer.js"></script>
```
Comment (abrugaro) [inline kai-client-ci/report/index.html:45]: `script` tags belong inside the `head`, if you want to load it after the html use the "defer" attribute 

```html
<script src="./visualizer.js" defer></script>
```

File: kai-client-ci/report/visualizer.js [hunk 3]
```diff
@@ -0,0 +1,500 @@
+let KAIData = []
+let JsonDates = []
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:2]: The variable names should start in lowercase

File: kai-client-ci/report/visualizer.js [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+            pieCharts(selectedDate);
+            averageRangesChart(selectedDate);
+        }
+    });
+}
+
+function formatDatesToLabels(dates) {
+
+    const months = ['Jan', 'Feb', 'March', 'April', 'May', 'June', 'July', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:72]: Missing semicolon, please take a look at the whole code to meet the code format PR check

File: kai-client-ci/report/visualizer.js [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+            JsonDates = KAIData.map(obj => obj.date.split(" ")[0]);
+            JsonDates.sort();
+            break;
+        } catch (error) {
+            console.error('There was a problem with the fetch operation:', error);
+            retries -= 1;
+            if (retries === 0) {
+                console.error('All retries failed. Please check your network connection or the URL.');
+            } else {
+                console.log(`Retrying... (${3 - retries} attempts left)`);
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:125]: The console logs can be removed

**Review by abrugaro — CHANGES_REQUESTED** [review]
Nice job!!! some minor comments

**Review by sshveta — COMMENTED** [review]
Looks Good . Consider separating the chart creation logic into smaller reusable functions that could generate different kinds of charts (e.g., bar, line, doughnut). This could improve maintainability.
Example : 


```
function createChart(ctx, chartType, chartData, chartOptions) {
    return new Chart(ctx, {
        type: chartType,        
        data: chartData,       
        options: chartOptions   
    });
}

function createLineChart(ctx, labels, data, colors) {
}
function createBarChart(ctx, labels, data, colors) {} etc
```

File: kai-client-ci/report/index.html (hunk 2 above)
Comment (midays) [reply kai-client-ci/report/index.html:45]: nice! done

File: kai-client-ci/report/visualizer.js (hunk 4 above)
Comment (midays) [reply kai-client-ci/report/visualizer.js:72]: done

File: kai-client-ci/report/visualizer.js (hunk 5 above)
Comment (midays) [reply kai-client-ci/report/visualizer.js:125]: done

**Comment by midays** [comment]
> Looks Good . Consider separating the chart creation logic into smaller reusable functions that could generate different kinds of charts (e.g., bar, line, doughnut). This could improve maintainability. Example :
> 
> ```
> function createChart(ctx, chartType, chartData, chartOptions) {
>     return new Chart(ctx, {
>         type: chartType,        
>         data: chartData,       
>         options: chartOptions   
>     });
> }
> 
> function createLineChart(ctx, labels, data, colors) {
> }
> function createBarChart(ctx, labels, data, colors) {} etc
> ```



> Looks Good . Consider separating the chart creation logic into smaller reusable functions that could generate different kinds of charts (e.g., bar, line, doughnut). This could improve maintainability. Example :
> 
> ```
> function createChart(ctx, chartType, chartData, chartOptions) {
>     return new Chart(ctx, {
>         type: chartType,        
>         data: chartData,       
>         options: chartOptions   
>     });
> }
> 
> function createLineChart(ctx, labels, data, colors) {
> }
> function createBarChart(ctx, labels, data, colors) {} etc
> ```

I've created a function called `createChart`

`function createChart(chartID, type, data, options) {
    const ctx = document.getElementById(chartID).getContext('2d');
    new Chart(ctx, {
        type: type,
        data: data,
        options: options
    });
}`

 as for the colors, and labels, the colors are statically defined at the top, I don't see why to pass them as an argument, and in regards of the lables they are considerd a part of the data object that I pass, so there's no need for it to have a separate variable, 

The only difference between a line and a bar chart ... etc. is just a string, so is it neccessary to have a separate function just for that? besides for the kaiPerformanceChart, it is made from 2 types of charts (line, and bar) so creating functions will just confuse more, WDYT ?

File: kai-client-ci/report/visualizer.js [hunk 6]
```diff
@@ -0,0 +1,497 @@
+let kaiData = []
+let jsonDates = []
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:2]: There are still some missing semicolons and other formatting issues that prevent the ci from passing.

File: kai-client-ci/report/visualizer.js [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+            document.getElementById('average-ranges-chart').remove();
+            const averageRangesCanvas = document.createElement('canvas');
+            averageRangesCanvas.id = 'average-ranges-chart';
+
+            document.getElementById('single-performace-charts').appendChild(validCodeCanvas);
+            document.getElementById('single-performace-charts').appendChild(averageRangesCanvas);
+
+            pieCharts(selectedDate);
+            averageRangesChart(selectedDate);
+        }
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:64]:
```suggestion
        if (!jsonDates.includes(selectedDate)) {
            return;
        }
        
        document.getElementById('pies-chart').remove();
        const validCodeCanvas = document.createElement('canvas');
        validCodeCanvas.id = 'pies-chart';

        document.getElementById('average-ranges-chart').remove();
        const averageRangesCanvas = document.createElement('canvas');
        averageRangesCanvas.id = 'average-ranges-chart';

        document.getElementById('single-performace-charts').appendChild(validCodeCanvas);
        document.getElementById('single-performace-charts').appendChild(averageRangesCanvas);

        pieCharts(selectedDate);
        averageRangesChart(selectedDate);
```

File: kai-client-ci/report/visualizer.js [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+                    return itemDate >= startDate && itemDate <= endDate;
+                });
+
+                document.getElementById('kai-performance-chart').remove();
+                const canvasContainer = document.createElement('canvas');
+                canvasContainer.id = 'kai-performance-chart';
+                document.getElementById('history-performance').appendChild(canvasContainer);
+
+                kaiPerformanceChart(filteredData);
+            }
```
Comment (abrugaro) [inline kai-client-ci/report/visualizer.js:100]:
```suggestion
            if (selectedDates.length !== 2) {
                return;
            }
            const startDate = selectedDates[0];
            const endDate = selectedDates[1];

            const filteredData = kaiData.filter(item => {
              const itemDate = new Date(item.date);
              return itemDate >= startDate && itemDate <= endDate;
            });

            document.getElementById('kai-performance-chart').remove();
            const canvasContainer = document.createElement('canvas');
            canvasContainer.id = 'kai-performance-chart';
            document.getElementById('history-performance').appendChild(canvasContainer);

            kaiPerformanceChart(filteredData);
```

**Review by abrugaro — CHANGES_REQUESTED** [review]
Please check the code format.

I left 2 nitpicks suggesstion that you can leave as they are as they're a matter of personal taste

File: kai-client-ci/report/visualizer.js (hunk 7 above)
Comment (midays) [reply kai-client-ci/report/visualizer.js]: nice suggestion!

**Comment by midays** [comment]
New PR: https://github.com/konveyor/kai-ci/pull/29

**Pull request closed without merging**